blog/content/post/2024-04-12-odroid-m1-secure-boot.md

```diff
@@ -0,0 +1,201 @@
+---
+title: Enabling Secure Boot on ODROID M1 (rk3568)
+abstract: 'Abstract first sentence.
+          Abstract second sentence.
+          Abstract third sentence.'
+cover: /covers/image-file.png
+author: michal.iwanicki
+layout: post
+published: true
+date: 2024-04-12
+archives: "2024"
+
+tags:
+  - bootloader
+  - secure-boot
+  - rockchip
+  - u-boot
+categories:
+  - Firmware
+  - Security
+
+---
+
+## Intro
+
+In this post I'll describe how to write key hash into OTP memory and enable
+pre-loader signature verification on ODROID-M1 which uses RK3568B2 SoC and also
+how to enable signature verification in U-Boot SPL.
+
+## Enabling Secure Mode
+
+### Preparation
+
+#### Repositories
+
+Writing key hash into OTP memory is done by U-Boot SPL which we will need to
+build. I used [Rockchip U-Boot](https://github.com/rockchip-linux/u-boot/)
+because it contains needed functionality implemented in
+`lib/rsa/rsa-verify.c`.
+[Hardkernel U-Boot](https://github.com/hardkernel/u-boot/tree/odroidm1-v2017.09)
+should also work. \
+We will also need [upgrade_tool](https://github.com/hardkernel/rk3568-linux-tools/tree/master/linux/Linux_Upgrade_Tool/Linux_Upgrade_Tool) and
+[rkbin](https://github.com/rockchip-linux/rkbin)
+
+#### Generating RSA Keys
+
+TODO
+
+#### Final directory structure
+
+```shell
+tree -FL 1
+./
+├── Linux_Upgrade_Tool/
+├── rkbin/
+├── u-boot/
+├── keys/
+└── upgrade_tool -> Linux_Upgrade_Tool/upgrade_tool
+```
+
+### Dependencies
+
+Below are packages needed to build U-Boot on Debian 11 OS.
+
+```shell
+apt install gcc make bison flex python3 python3-dev python3-setuptools python3-pyelftools swig libssl-dev device-tree-compiler python2 bc
+```
+
+To build Rockchip U-Boot without errors we will also need
+[Linaro gcc aarch64 toolchain](https://releases.linaro.org/components/toolchain/binaries/6.3-2017.05/aarch64-linux-gnu/).
+
+### Configuration
+
+#### rkbin
+
+In `rkbin/RKBOOT/RK3568MINIALL.ini` update FlashBoot so it points to
+`u-boot-spl.bin` file
+
+```diff
+diff --git a/RKBOOT/RK3568MINIALL.ini b/RKBOOT/RK3568MINIALL.ini
+index ef6d071bc5ac..c80d817f4b94 100644
+--- a/RKBOOT/RK3568MINIALL.ini
++++ b/RKBOOT/RK3568MINIALL.ini
+@@ -15,7 +15,7 @@ NUM=2
+ LOADER1=FlashData
+ LOADER2=FlashBoot
+ FlashData=bin/rk35/rk3568_ddr_1560MHz_v1.21.bin
+-FlashBoot=bin/rk35/rk356x_spl_v1.13.bin
++FlashBoot=../u-boot/spl/u-boot-spl.bin
+ [OUTPUT]
+ PATH=rk356x_spl_loader_v1.21.113.bin
+ [SYSTEM]
+```
+
+#### U-Boot
+
+Update `CROSS_COMPILE_ARM64` variable inside `make.sh` with correct path to
```

Comment:
@m-iwanicki please use I/We instead of you. We should write those posts in first-person form

Reply:
@TomaszAIR New commit is here:  af72329. I hope I managed to use I/We everywhere.